### PR TITLE
First attempt to implement Majority-honest multiplication protocol

### PR DESCRIPTION
**Pull request by akoshelev**
Based on [this paper](https://eprint.iacr.org/2019/1390.pdf), this PR adds a `shmc` module that contains components that participate in semi-honest multiplication.

### CODE REVIEW

**Comment by akoshelev** [comment]
fixed cargo issues detected [here](https://github.com/martinthomson/raw-ipa/runs/6103108539?check_suite_focus=true)

File: src/shmc/mod.rs [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+    }
+
+    /// Shares this secret to 3 parties by choosing three random elements (v1..v3)
+    /// under the constraint that v1+v2+v3 == self
+    #[must_use]
+    pub fn share(&self) -> [Share; 3] {
+        // thread_rng() is secure, but may not be secure enough for us. need to check
+        // what rand crate uses for it.
+        let u1 = thread_rng().gen::<Self>();
+        let u2 = thread_rng().gen::<Self>();
```
Comment (martinthomson) [inline src/shmc/mod.rs:108]: Please don't internalize the RNG.  This is something that should be provided as an argument to this function.  I would also add an API that allows the caller to specify the random values explicitly, which can be invaluable in testing.  That means a signature like:

```rust
// type Rng = rand::Rng + rand::CryptoRng + ?Sized;
fn share(&self, rng: &mut impl Rng) -> [Share; 3] {
  self.share_with(rng.gen::<Self>(), rng.gen::<Self>())
}
fn share_with(&self, r1: Self::Type, r2: Self::Type) -> [Share; 3] {
  let r3 = self - r1 - r2;
  [Share(r1, r3), Share(r2, r1), Share(r3, r2)]
}
```

**Review by martinthomson — COMMENTED** [review]
Some high-level stuff.  Generally this seems to be aimed in the right direction.

File: src/shmc/mod.rs [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+use hkdf::Hkdf;
+use rand::distributions::{Distribution, Standard};
+use rand::{thread_rng, Rng};
+use sha2::Sha256;
+
+/// Represents a secret in the system that must obey the following properties:
+/// - addition
+/// - multiplication
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Debug)]
+pub struct SharedSecret(u128);
```
[the remark this answers is not in the document]
Comment (akoshelev) [reply src/shmc/mod.rs:50]: Yea, `Scalar` seems to be what I wanted here - thanks for pointing that out! I am just going to get rid of this type

File: src/shmc/mod.rs [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+    /// use raw_ipa::shmc::SharedSecret;
+    /// let u = SharedSecret::from(5);
+    /// let v = SharedSecret::from(10);
+    ///
+    /// assert_eq!(SharedSecret::from(15), u.wrapping_add(v));
+    /// ```
+    #[must_use]
+    pub fn wrapping_add(self, other: Self) -> Self {
+        SharedSecret(self.0.wrapping_add(other.0))
+    }
```
[the remark this answers is not in the document]
Comment (akoshelev) [reply src/shmc/mod.rs:71]: make sense, I wasn't sure about adding third-party crates, but given that `Scalar` has everything what I want here, I am just getting rid of this struct + impl

File: src/shmc/mod.rs (hunk 1 above)
Comment (akoshelev) [reply src/shmc/mod.rs:108]: +1 thanks! this is strictly better

File: src/shmc/mod.rs [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+        let u1 = shares[0].0;
+        let u2 = shares[1].0;
+        let u3 = shares[2].0;
+
+        u1.wrapping_add(u2).wrapping_add(u3)
+    }
+
+    /// Computes KDF over this secret with a given seed.
+    #[must_use]
+    pub(crate) fn kdf(&self, seed: u64) -> Self {
```
[the remark this answers is not in the document]
Comment (akoshelev) [reply src/shmc/mod.rs:134]: yep I was aware of it, had to use something to make the whole thing work. Given that I have `Scalar` now, will use it instead

File: src/shmc/mod.rs [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+/// A share of `SharedSecret`.
+#[derive(Copy, Clone, Debug)]
+pub struct Share(SharedSecret, SharedSecret);
+
+impl Share {
+    /// Computes HKDF over this share and returns a new instance of `SharedSecret`.
+    #[must_use]
+    pub fn kdf(&self, seed: u64) -> SharedSecret {
+        self.0.kdf(seed).wrapping_sub(self.1.kdf(seed))
+    }
```
[the remark this answers is not in the document]
Comment (akoshelev) [reply src/shmc/mod.rs:168]: yea hopefully this concern goes away with us using 32 byte `Scalar` for seed/secret

File: src/shmc/mod.rs [hunk 6]
```diff
@@ -0,0 +1,508 @@
...
+impl Distribution<SharedSecret> for Standard {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> SharedSecret {
+        let v = rng.gen();
+
+        SharedSecret(v)
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct Randomness {
```
[the remark this answers is not in the document]
Comment (akoshelev) [reply src/shmc/mod.rs:154]: I am not sure I understand the suggestion here - multiplication is a stateful process managed by `Helper`. `Randomness` is just a container to transfer random bits with seed to a `Helper`. As mentioned in the module overview, we deviated from paper's implementation here and instead, for now, we generate random secret + seed on the client side with the same invariant that (`a1` + `a2` + `a3` == 0)

**Comment by akoshelev** [comment]
closing it as we no longer need this implementation

**Pull request closed without merging**